SynCommons.pas

```diff
@@ -16763,14 +16769,49 @@ implementation
   UTF8_FIRSTBYTE: array[2..6] of byte = ($c0,$e0,$f0,$f8,$fc);
 
 {$ifdef FPC}
+
+{$if not defined (CPUARM3264) and not defined(BSD)}
 procedure VarClear(var v: variant); // defined here for proper inlining
+  (*
+  Alf.
+  This has been a very long standing issue on all arm/arm64 and BSD systems.
+  The sockets got corrupted by the variants test, leading to TestSQL3 failures.
+  Been hunting this for more than 2 years ...
+  It should/ could be investigated in more detail.
+  Also due to the fact that, for FPC this VTYPE_STATIC is often disabled in the source by defines !
+  *)
 var p: pointer; // more efficient generated asm with an explicit temp variable
 begin
   p := @v;
   if integer(PVarData(p)^.VType) and VTYPE_STATIC=0 then

```

Comment:
Did you try to get rid of integer() here ?
for instance, with:
_if word(PVarData(p)^.VType) and VTYPE_STATIC = 0 then_

Perhaps it may be a nicer fix.

Reply:
See my comment below.
I have no clue about this memory corruption. It should not happen. Its a compiler thing I guess.
A mystery for me. And I am not going to dive into the compiler.